### PR TITLE
Bugfix/mirror all patch conflicts

### PR DESCRIPTION
**Pull request by Paul-Ferrell**
Closes https://github.com/spack/spack/pull/13622

 - Appends part of the patch has to the patch name to differentiate
   patches with the same name (for the same package).
 - Patches are re-downloaded if the file isn't at the storage path
   or it's cosmetic name are missing.
 - Fixed an exception handling error in mirror.py (not all exceptions)
   have a 'message' attribute.

### CODE REVIEW

File: var/spack/repos/builtin/packages/mpfr/package.py [hunk 1]
```diff
@@ -39,8 +39,10 @@ class Mpfr(AutotoolsPackage):
     }
 
     for ver, checksum in patches.items():
-        patch('https://www.mpfr.org/mpfr-{0}/allpatches'.format(ver),
-              when='@' + ver, sha256=checksum)
+        patch(
```
Comment (scheibelp) [inline var/spack/repos/builtin/packages/mpfr/package.py]: This looks the same but with a different style - if I'm right can you revert the changes to `mpfr`?

File: lib/spack/spack/patch.py [hunk 2]
```diff
@@ -199,11 +199,15 @@ def fetch(self, stage):
         fetcher = fs.URLFetchStrategy(self.url, fetch_digest,
                                       expand=bool(self.archive_sha256))
 
-        per_package_ref = os.path.join(
-            self.owner.split('.')[-1], os.path.basename(self.url))
+        # Patch url's can be ambiguous - The same patch name can have different
+        # versions that apply to different versions of the package. We append
+        # a bit of the hash to differentiate them. (Collision odds are exceedingly low
+        # given how few of these there should be, including b-day attack odds). 
+        name = '{0}-{1}'.format(os.path.basename(self.url), fetch_digest[:8])
```
Comment (scheibelp) [inline lib/spack/spack/patch.py]: (very minor) where we use a subset of the hash we typically use the first 7 characters - can you change "8" to "7"?

Also I think the comment could be improved. Perhaps

> The same package can have multiple patches with the same name but with different contents, therefore apply a subset of the hash

also I don't think the qualification about collision is necessary (we do this in a few other spots without mentioning it), although I agree with your reasoning.

File: lib/spack/spack/stage.py [hunk 3]
```diff
@@ -498,7 +498,8 @@ def cache_mirror(self, stats):
         absolute_storage_path = os.path.join(
             dst_root, self.mirror_paths.storage_path)
 
-        if os.path.exists(absolute_storage_path):
+        if (os.path.exists(absolute_storage_path) and
```
Comment (scheibelp) [inline lib/spack/spack/stage.py]: If the storage path exists and the cosmetic path doesn't yet exist, then generally the only requirement is that we need to create the symlink. As is, this will re-download a file just to recreate the symlink. I realize that I didn't make this easy to accommodate but I'm wondering whether either (a) we can figure out how to just recreate the symlink here or (b) this can be moved to another PR.

File: lib/spack/spack/patch.py (hunk 2 above)
Comment (Paul-Ferrell) [reply lib/spack/spack/patch.py]: Done.

File: lib/spack/spack/stage.py (hunk 3 above)
Comment (Paul-Ferrell) [reply lib/spack/spack/stage.py]: The fetch step happens before this, and regardless of whether these paths exist.
If we remove this it means mirrors won't fix the cosmetic paths unless we go in and delete the files at the storage paths. That's especially problematic in this case, as this whole issue means that we've lost the links to many of those storage paths in the first place.

Comment (scheibelp) [reply lib/spack/spack/stage.py]: There's a call to `fetch` 5 lines down from here, that's where I intend to do the fetch; I intend that if the `storage_path` is already stored that we skip the fetch. Where do you observe that the fetch occurs before the call to `cache_mirror` (it is a mistake on my part that a fetch is occurring before this, is why I ask).

Comment (Paul-Ferrell) [reply lib/spack/spack/stage.py]: Line 508 in mirror.py calls patch.fetch, which generates a stage just for the patch url and fetches it (on line 214 in patch.py).

Comment (Paul-Ferrell) [reply lib/spack/spack/stage.py]: Also, I can't believe I missed that fetch in cache_mirror. It was, as you said, right there.

Comment (scheibelp) [reply lib/spack/spack/stage.py]: Ah ok so there's a difference in how patches are handled vs. other resources. I think the call to `patch.fetch` in mirror can actually be removed, since that will be handled in the `cache_mirror` call. I think the easiest way to handle this would be to separate out the symlink creation in `MirrorCache` into a separate function, and have `Stage.cache_mirror` invoke this function regardless of whether the `storage_path` already exists. If you like that could be handled in a separate PR (by me or you, according to your preference); here works too.

Comment (Paul-Ferrell) [reply lib/spack/spack/stage.py]: I did the separation in this merge request. I left the `patch.fetch` in mirror alone though.

**Comment by scheibelp** [comment]
Thanks!
